### ISSUE TITLE
Make sure Session is correctly used.

### ISSUE THREAD

**Issue by chrhol** [Starcounter/Starcounter#3558]
A `Session` is bound to a scheduler, but verification that it's used from the correct one is only implemented in a few places, which might lead to undefined behaviour. Also since we use a lock to assure that only one thread is accessing a session at a time, we need to secure that no state is accessed without first getting the lock.

- `Session.Use` should check that it's called from the same scheduler as it was created in.
- If not on the correct scheduler or if it should be accessed async, a task needs to be scheduled using `Session.ScheduleTask`.
- All public properties in Session should verify that the session is the current running one (if not it needs to be scoped with `Session.Use` or `Session.ScheduleTask`)

These changes might lead to that current applications will break, since it's possible to access the session incorrectly in the current versions of Starcounter. 

/cc: @dan31, @malx122, @miyconst, @warpech, @mtornwall, @ripipip

**Comment by mtornwall** [comment]
> These changes might lead to that current applications will break, since it's possible to access the session incorrectly in the current versions of Starcounter.

Well, based on the bug description it sounds like such applications are already broken. :-)

I'll take predictable crashing over UB any day of the week.

**Comment by miyconst** [comment]
>Well, based on the bug description it sounds like such applications are already broken. :-)

+1

As of now I can't remember any application which would use Starcounter bug to hijack session of another app and communicate between each other that way. When needed to access another session we use `Session.ForAll` or `Session.ScheduleTask`.

I still don't understand whether, when, and how `Session.Use` should be used inside applications.

**Comment by chrhol** [comment]
>  I still don't understand whether, when, and how Session.Use should be used inside applications.

Yeah, maybe we should remove that one and always use `ScheduleTask` instead.

**Comment by mtornwall** [comment]
> I still don't understand whether, when, and how Session.Use should be used inside applications.

I use this pattern (and its predecessors `StartUsing` and `StopUsing`) extensively to receive the results of asynchronous background jobs. If I can do that using `ScheduleTask` instead I have no objections to the removal of `Use`.

**Comment by aloscha** [comment]
I serveral apps we use Session.Current.Data, but now that code return null value.
I do small investigation and create demo to show the problem https://github.com/warpech/HelloWorld/tree/Session_example (that branch from small hello world app).

Generally I register in main application (HelloWorld) PersonJson and InjectionJson. In second app (HelloWorldPart) I register just handler to get from previous app InjectionJson. So it's working that (please run first HelloWorld and HelloWorldPart):
1. Go to HelloWorld
2. Page Master set current page to PersonJson
3. Meanwhile other app is in ontology maping to Illustration, which called with PersonJson
4. HelloWorldPart called helloworldpart, which is ontology maping to Somebody
5. InjectionJson is registered to Somebody so content is send to HelloWorldPart and then send back to HelloWorld

(Yeah it looks a bit confused)

So after click to Announce buttom, we will have exception in InjectionJson because Session.Current.Data is null.
![image](https://cloud.githubusercontent.com/assets/5141943/15109577/4e653510-15dc-11e6-8150-e25f4830ad63.png)

Please notice that in older version that demo works, I mean untill 2.2.1936 is released (so in 2.2.1901 (29 April 2016 04:56:23) it works, but start from 2.2.1936 (30 April2016 04:57:11) don't)

So my question is how to use Session in proper way since 2.2.1936 is released (or it's some bug)?

**Comment by chrhol** [comment]
Sorry for not getting back to you sooner @aloscha. The `HelloWorldPart` folder and project seems to be missing from this branch. It's included in the solution but fails to load. 

There have been no changes directly for Sessions between these versions what I know of, only things around it, which would point to that it's a bug somewhere.

**Comment by aloscha** [comment]
Sorry @chrhol, I create Part solution in separate folder. Now updated main HelloWorld branch, by adding that project.

**Comment by chrhol** [comment]
Something is not working as it should. I will move the discussion to a separate issue and continue there.

**Comment by chrhol** [comment]
Most of the changes for Session are in place. Now access to the session should be properly verififed that it's used from the correct scheduler, that access is obtained and that it's not destroyed.

I will spend some time testing it with external apps today to make sure it works as intended. I will push the changes later.

**Comment by chrhol** [comment]
Tested and found a couple if things that needed to be fixed. I will hold off pushing it to develop/pnext though until the next stable version is published to avoid introducing anything unforeseen. 

The changes can be found in branch https://github.com/Starcounter/Starcounter/tree/issue/3558

**Comment by dan31** [comment]
Current 2.2.1 Stable candidate is now (finally) being staged.
